### PR TITLE
fix!: prevent signature verification from resolving against Docker Hub

### LINKED ISSUE #767: cache/verify: Signature verification resolves OCI references against Docker Hub instead of configured registry
----
> 🦾 Written with LLM assistance [claude-opus-4-6]
> 💪 Reviewed by a human before submission
----

**Describe the bug**

Signature verification resolves OCI artifact references against Docker Hub (`index.docker.io`) instead of the registry configured in `complytime.yaml`. This happens because `BuildLookupRef()` in `internal/cache/sync.go` constructs a reference without a registry host (e.g. `policies/verify-keyed:v1.0.0`), and when this is passed to `name.ParseReference()` from `go-containerregistry` at `internal/cache/verify.go:216`, it defaults to `index.docker.io`.

This affects both policy and complypack verification workflows.

**To Reproduce**

1. Configure `complytime.yaml` with a policy hosted on a private OCI registry (e.g. `registry.example.com/policies/my-policy`)
2. Add a `verification:` block with a public key or keyless identity
3. Run `complyctl get`
4. Observe that the signature verification lookup hits `index.docker.io` instead of `registry.example.com`

**Expected behavior**

Signature verification should resolve the OCI reference against the same registry that the policy was fetched from. The registry host from the policy entry should be threaded through to the verification function so `name.ParseReference()` receives a fully-qualified reference (e.g. `registry.example.com/policies/my-policy:v1.0.0`).

**Additional context**

### Affected code paths

All verification call sites use `BuildLookupRef()` which constructs a registry-less reference:

1. **Policy sync** — `internal/cache/sync.go:116`:
   ```go
   registryRef := BuildLookupRef(policyID, tag, digest)
   ```
   `policyID` (e.g. `policies/verify-keyed`) has no registry host.

2. **Complypack sync** — `internal/cache/complypack_sync.go:139`:
   ```go
   registryRef := BuildLookupRef(repository, tag, digest)
   ```
   Same issue — `repository` lacks a registry host.

3. **Complypack local cache re-verification** — `internal/cache/complypack_sync.go:218`:
   ```go
   registryRef := BuildLookupRef(repository, tag, digest)
   ```

4. **Reference parsing** — `internal/cache/verify.go:216`:
   ```go
   ref, err := name.ParseReference(registryRef)
   ```
   `name.ParseReference()` defaults to `index.docker.io` when no host is present.

### Root cause

`BuildLookupRef()` (`sync.go:20-28`) constructs the reference from `policyID`/`repository`, `tag`, and `digest` without a registry host component. Neither the `Sync` struct nor the `ComplypackSync` struct carries a `registryHost` field that could be used to qualify the reference.

### Suggested fix

Thread the registry host from the policy entry's OCI reference through to `SyncPolicy()`/`SyncComplypack()`. Options:

- Add a `registryHost` field to `Sync` and `ComplypackSync` structs, populated from the policy entry's registry reference during construction
- Modify `BuildLookupRef()` to accept a registry host parameter
- Extract the host from the source reference already available in the sync context

### Impact

- Blocks #678 since it needs to use a local registry for evaluation

### PR DESCRIPTION
**Pull request by trevor-vaughan**
Fixes signature verification resolving against Docker Hub instead of the
configured registry when verifying policies and complypacks.
`BuildLookupRef()` now accepts the registry host as its first parameter
and prepends it to OCI references passed to `VerifyFunc`. When a
verifier is configured but the registry host is empty, sync operations
fail closed with an actionable error message directing users to include
the registry host in the policy URL or use `--skip-verify`.

This is a security fix addressing CWE-940 (Improper Verification of
Source of a Communication Channel)

Closes #767

**Core signature verification fix:**
  - `BuildLookupRef()` gains `registryHost` parameter prepended to OCI references
  - `Sync` and `ComplypackSync` structs gain `registryHost` field populated from `ref.Registry`
  - `NewSync()` and `NewComplypackSync()` constructors gain `registryHost` parameter
  - All `VerifyFunc` invocations receive host-qualified references

**Fail-closed validation:**
  - Returns error when verifier configured but `registryHost` is empty
  - Prevents silent misresolution to Docker Hub
  - Error messages include remediation guidance (add registry host or use `--skip-verify`)

**Bug fixes:**
  - `DefinitionVersion` call sites pass empty `registryHost` (resolves host internally via `Client.buildRef()`)
  - Prevents double-host references like `ghcr.io/ghcr.io/org/policy`

**Documentation:**
  - AGENTS.md Recent Changes entry documents BREAKING signature change
  - Threat model THR02 refined: MIT02 notes host-qualified verification, MIT04 added for fail-closed behavior

**Testing:**
  - `TestSync_RegistryHost_PassedToVerifier` captures `registryRef` and asserts configured host prefix
  - `TestSync_EmptyRegistryHost_FailClosed` asserts error when verifier set without host
  - Four new `BuildLookupRef` test cases (host+tag, host+digest, host+empty, host+latest)

- The change is breaking:
  - `BuildLookupRef()`, `NewSync()`, and `NewComplypackSync()` signatures changed
  - Users who relied on the default docker fallback will now receive an error
- Review the fail-closed behavior in `sync.go:98-103` and `complypack_sync.go:115-120`
- OpenSpec artifacts in `openspec/changes/registry-host-verification-fix/` provide full requirements and design rationale
- All existing callers updated; backward compatibility maintained via empty string parameter for non-verified flows

Assisted-by: LLM (claude-sonnet-4-5)
Signed-off-by: Trevor Vaughan <tvaughan@redhat.com>

### CODE REVIEW

**Review by marcusburghardt — COMMENTED** [review]
Overall this is a clean, well-structured security fix for CWE-940 with comprehensive test coverage and spec traceability. Two non-blocking observations below for the author's consideration.

_This review was generated by /review-pr (AI-assisted)._

File: internal/cache/sync.go [hunk 1]
```diff
@@ -116,7 +129,12 @@ func (s *Sync) SyncPolicy(ctx context.Context, policyID, version string) (bool,
 	// content to disk. If verification fails, the local cache is unchanged.
 	var verifyResult *VerificationResult
 	if s.verifier != nil {
-		registryRef := BuildLookupRef(policyID, tag, digest)
+		// Fail-closed: refuse to verify against an unqualified reference that
+		// would silently resolve to Docker Hub instead of the intended registry.
+		if s.registryHost == "" {
```
Comment (marcusburghardt) [inline internal/cache/sync.go:134]: **[LOW] Fail-closed check placement differs from complypack path**

This check is inside the verification block, which executes only when a fresh copy is needed (after `DefinitionVersion` returns and the incremental skip check passes). In `complypack_sync.go`, the same check is at the entry of `SyncComplypack` before any network call.

Two practical differences:

1. A misconfigured setup (verifier set, no registry host) won't surface the error for policies that are already cached — only when the policy version changes upstream and a re-fetch is needed. The complypack path catches it on every invocation.
2. On cache misses, a `DefinitionVersion` network round-trip completes before the error is returned. Moving the check before `DefinitionVersion` (matching the complypack pattern) would fail faster with no wasted network call.

Not blocking since `ValidateOCIRef` at config time provides the primary guard, but harmonizing the placement would improve consistency and UX.

File: openspec/config.yaml [hunk 2]
```diff
@@ -0,0 +1,20 @@
+schema: spec-driven
```
Comment (marcusburghardt) [inline openspec/config.yaml:1]: **[LOW] Is this file intentionally included?**

This file contains only the `schema: spec-driven` declaration and commented-out examples. The change directory already declares its own schema in `.openspec.yaml`.

Was this inclusion intentional (e.g., adopting OpenSpec root config as a project convention)? If not, it could be removed from this PR to keep the scope focused on the security fix. If the file isn't needed going forward, consider adding it to `.gitignore`.

**Review by marcusburghardt — APPROVED** [review]
Approving -- this is a clean, well-structured security fix for CWE-940 with comprehensive test coverage, spec traceability, and all CI checks passing.

Two LOW-severity follow-ups filed as draft PR #776 (on top of this branch):
1. Move the fail-closed check in `SyncPolicy` before `DefinitionVersion` to match the `ComplypackSync` pattern (avoids a wasted network call on misconfigured setups)
2. Remove the auto-scaffolded `openspec/config.yaml` boilerplate to keep scope focused

Neither is blocking.

_This review was generated by /review-pr (AI-assisted)._

File: internal/cache/sync_test.go [hunk 3]
```diff
@@ -377,21 +377,28 @@ func TestSync_ConcurrentMixedFailures(t *testing.T) {
...
+		{"tag version", "", "org/policy", "v1.0", "", "org/policy:v1.0"},
+		{"sha256 digest", "", "org/policy", "", "sha256:9f86d081884c7d659a2feaa0c55ad015a3bf4f1b2b0b822cd15d6c15b0f00a08", "org/policy@sha256:9f86d081884c7d659a2feaa0c55ad015a3bf4f1b2b0b822cd15d6c15b0f00a08"},
+		{"sha512 digest", "", "org/policy", "", "sha512:def456", "org/policy@sha512:def456"},
+		{"empty version", "", "org/policy", "", "", "org/policy"},
+		{"latest version", "", "org/policy", "latest", "", "org/policy"},
+		// New cases: non-empty registryHost prepends host prefix.
+		{"host with tag", "registry.example.com", "policies/verify-keyed", "v1.0.0", "", "registry.example.com/policies/verify-keyed:v1.0.0"},
+		{"host with digest", "registry.example.com", "policies/verify-keyed", "", "sha256:abc123", "registry.example.com/policies/verify-keyed@sha256:abc123"},
+		{"host with empty version", "registry.example.com", "org/policy", "", "", "registry.example.com/org/policy"},
+		{"host with latest", "registry.example.com", "org/policy", "latest", "", "registry.example.com/org/policy"},
```
Comment (em-redhat) [inline internal/cache/sync_test.go:397]: **[HIGH]** Missing edge cases for `BuildLookupRef` boundary inputs:

1. **Port-based registryHost**: Port-based registries (`localhost:5000`, `registry.internal:5001`) are a common deployment pattern. Add a table entry:
```go
{"host with port and tag", "myregistry.io:5000", "org/policy", "v1.0.0", "", "myregistry.io:5000/org/policy:v1.0.0"},
```

2. **Trailing slash**: `BuildLookupRef` concatenates `registryHost + "/" + repository`. A trailing slash in `registryHost` would produce `registry.example.com//org/policy`. While `ref.Registry` from `ParsePolicyRef()` likely strips trailing slashes, the unit test should document the function's contract for this input. Either add a test case or add `strings.TrimRight(registryHost, "/")` normalization.

File: openspec/changes/registry-host-verification-fix/tasks.md [hunk 4]
```diff
@@ -0,0 +1,37 @@
+## 1. Update BuildLookupRef signature and tests (TDD: test first)
+
+- [ ] 1.1 Update `TestBuildLookupRef` table in `internal/cache/sync_test.go`: add `registryHost` field to test struct, keep existing cases with empty host for backward compatibility, add new cases with non-empty host (tests will fail to compile)
```
Comment (em-redhat) [inline openspec/changes/registry-host-verification-fix/tasks.md:3]: **[HIGH]** All task checkboxes remain `[ ]` despite implementation being complete. Per AGENTS.md Behavioral Rules: "Mark checkboxes `[x]` immediately on completion." Mark all completed tasks as `[x]`.

File: internal/cache/complypack_sync_test.go [hunk 5]
```diff
@@ -891,3 +894,120 @@ func TestComplypackSync_VPrefixedTag_StateMatchesDisk(t *testing.T) {
...
+	require.NoError(t, err)
+	assert.True(t, fetched3, "local cache hit should return true")
+
+	// All captured refs must have the registry host prefix.
+	mu.Lock()
+	refs := make([]string, len(capturedRefs))
+	copy(refs, capturedRefs)
+	mu.Unlock()
+
+	require.GreaterOrEqual(t, len(refs), 3,
```
Comment (em-redhat) [inline internal/cache/complypack_sync_test.go:975]: **[MEDIUM]** `GreaterOrEqual(t, len(refs), 3)` would pass if the verifier were called 4, 5, or 100 times -- potentially masking bugs where verification is called redundantly. The test comment says "pre-copy x2 + cache re-verify x1" which implies exactly 3 calls. Consider:
```go
assert.Equal(t, 3, len(refs), "verifier should be called exactly 3 times (pre-copy x2 + cache re-verify x1)")
```
If the count is genuinely variable, document why.

File: internal/cache/sync_test.go [hunk 6]
```diff
@@ -546,3 +555,68 @@ func TestSync_NilVerifier_SkipsVerification(t *testing.T) {
...
+	state, err := cache.LoadState(cacheDir)
+	require.NoError(t, err)
+
+	sync := cache.NewSync(cacheMgr, state, mock, cacheDir, "registry.example.com",
+		cache.WithVerifier(capturingVerifier))
+
+	_, err = sync.SyncPolicy(context.Background(), "org/test-policy", "latest")
+	require.NoError(t, err)
+
+	assert.True(t, strings.HasPrefix(capturedRef, "registry.example.com/"),
```
Comment (em-redhat) [inline internal/cache/sync_test.go:589]: **[MEDIUM]** No integration-level test covers the full `get.go` -> `ref.Registry` -> `NewSync` -> verifier path. The unit tests verify `BuildLookupRef` and `SyncPolicy` in isolation, but nothing tests that `ParsePolicyRef("ghcr.io/org/policy:v1.0.0").Registry` produces `"ghcr.io"` and that this value correctly flows through the constructor to the verifier. The existing E2E/acceptance tests (which pass) cover this implicitly, but an explicit unit-level test in `get_verify_test.go` would strengthen regression protection. Non-blocking -- noting for consideration.

File: internal/cache/complypack_sync_test.go [hunk 7]
```diff
@@ -891,3 +894,120 @@ func TestComplypackSync_VPrefixedTag_StateMatchesDisk(t *testing.T) {
...
+	evalID := "io.complytime.opa"
+	host := "registry.example.com"
+
+	// Seed with both host-qualified ref (for DefinitionVersion lookups) and
+	// bare repository (for CopyComplypack lookups).
+	mock.seedComplypack(host+"/"+repository, evalID, "1.0.0", "sha256:host-v1", "v1 content")
+	mock.seedComplypack(repository, evalID, "1.0.0", "sha256:host-v1", "v1 content")
+
+	// Track all refs passed to the verifier across call sites.
+	var mu sync.Mutex
```
Comment (em-redhat) [inline internal/cache/complypack_sync_test.go:921]: **[MEDIUM]** This `sync.Mutex` guards `capturedRefs` but all three sync calls below (lines 937, 951, 965) are sequential -- no concurrent goroutine access. The mutex adds cognitive overhead without functional benefit. Consider removing it to simplify the test. If `SyncComplypack` could internally invoke the verifier from a goroutine, add a comment explaining why the mutex is needed.

**Review by em-redhat — COMMENTED** [review]
## Review Council Report

**Agents**: Architect, Adversary (Security), Testing, Guard (Intent Drift), SRE, Scribe (Documentation)

| Agent | Verdict | Critical | High | Medium | Low |
|-------|---------|----------|------|--------|-----|
| Architect | APPROVE | 0 | 0 | 2 | 3 |
| Adversary | APPROVE | 0 | 0 | 0 | 2 |
| Testing | REQUEST CHANGES | 0 | 2 | 2 | 2 |
| Guard | APPROVE | 0 | 0 | 1 | 2 |
| SRE | APPROVE | 0 | 0 | 0 | 3 |
| Scribe | APPROVE | 0 | 1 | 3 | 3 |

### Security Assessment: STRONG

The Adversary verified all 3 `VerifyFunc` call sites use `BuildLookupRef(s.registryHost, ...)`, `registryHost` is immutable after construction (no TOCTOU), `ValidateOCIRef` provides config-time defense-in-depth, and fail-closed prevents silent Docker Hub fallback. No bypass vectors found. No new dependencies.

### Verdict: APPROVE (conditional on HIGH findings in inline comments)

The core security fix is well-designed and well-tested. The `registryHost` threading from `ref.Registry` through constructors to `BuildLookupRef` with fail-closed guards is correct. See inline comments for required changes and suggestions.

File: internal/cache/sync_test.go (hunk 3 above)
Comment (marcusburghardt) [reply internal/cache/sync_test.go:397]: Addressed in #776 ([`145c11b2`](https://github.com/complytime/complyctl/pull/776/commits/145c11b2)).

Both points covered:

1. **Port-based registries**: Added two test cases -- `myregistry.io:5000` with tag and `localhost:5000` with digest.
2. **Trailing slash**: Added `strings.TrimRight(registryHost, "/")` normalization in `BuildLookupRef` plus a test case (`"registry.example.com/"` -> `"registry.example.com/org/policy:v1.0.0"`).

@em-redhat would you be OK approving #771 as-is and letting us land these improvements via #776 (stacked on top of this branch)? That way the security fix is unblocked and the review findings ship right after.

File: openspec/changes/registry-host-verification-fix/tasks.md (hunk 4 above)
Comment (marcusburghardt) [reply openspec/changes/registry-host-verification-fix/tasks.md:3]: Addressed in #776 ([`145c11b2`](https://github.com/complytime/complyctl/pull/776/commits/145c11b2)). All 18 task items marked `[x]`.

File: internal/cache/complypack_sync_test.go (hunk 5 above)
Comment (marcusburghardt) [reply internal/cache/complypack_sync_test.go:975]: Addressed in #776 ([`145c11b2`](https://github.com/complytime/complyctl/pull/776/commits/145c11b2)). Changed to `require.Equal(t, 3, len(capturedRefs))` -- the test exercises exactly 3 call sites (pre-copy x2 + cache re-verify x1), so a loose bound would mask redundant verification bugs.

File: internal/cache/sync_test.go (hunk 6 above)
Comment (marcusburghardt) [reply internal/cache/sync_test.go:589]: Agreed this is non-blocking. E2E and acceptance tests (both passing) cover the full `get.go` -> `ref.Registry` -> verifier flow implicitly. Not included in #776 to keep the follow-up scoped to the actionable findings.

File: internal/cache/sync.go (hunk 1 above)
Comment (marcusburghardt) [reply internal/cache/sync.go:134]: Addressed in #776

File: openspec/config.yaml (hunk 2 above)
Comment (marcusburghardt) [reply openspec/config.yaml:1]: Addressed in #776

File: internal/cache/complypack_sync_test.go (hunk 7 above)
Comment (marcusburghardt) [reply internal/cache/complypack_sync_test.go:921]: Addressed in #776 ([`145c11b2`](https://github.com/complytime/complyctl/pull/776/commits/145c11b2)). Removed the mutex and added a comment noting the sequential nature of the calls. All three `SyncComplypack` invocations are sequential -- no concurrent goroutine access.

**Review by em-redhat — APPROVED** [review]
Comments addressed in #776
